src/main/java/com/google/maps/GeolocationApi.java

```diff
@@ -44,27 +47,56 @@ private GeolocationApi () {
   }
 
   public static PendingResult<GeolocationResult> geolocate(GeoApiContext context, GeolocationPayload payload) {
+    return context.post(GEOLOCATION_API_CONFIG, Response.class, payload);
+  }
 
-    return context.post(GEOLOCATION_API_CONFIG, GeolocationResponse.class, payload);
+  public static GeolocationApiRequest newRequest(GeoApiContext context) {
+    return new GeolocationApiRequest(context);
   }
 
-  private static class GeolocationResponse implements ApiResponse<GeolocationResult> {
-    private GeolocationResult result;
-    private ApiError error;
+   public static class Response implements ApiResponse<GeolocationResult> {
+     public int code = 200;
+     public String message = "OK";
+     public double accuracy = -1.0;
+     public LatLng location = null;
+     public String domain = null;
+     public String reason = null;
+     public String debugInfo = null;
 
     @Override
     public boolean successful() {
-      return error == null;
+      return code == 200;
     }
-
     @Override
     public GeolocationResult getResult() {
+      GeolocationResult result = new GeolocationResult();
+      result.accuracy = accuracy;
+      result.location = location;
       return result;
     }
-
     @Override
     public ApiException getError() {
-      return ApiException.from(error.status, error.message);
+      if (successful()) {
+        return null;
+      }
+      ApiException e;
+      // try and fit the older error codes into the new style geo api error formats
+      if(reason.equals("keyInvalid")) {
```

Comment:
Can you move these into the ApiException.from method, where the rest of the string-to-exception mapping occurs?